clientconn.go

```diff
@@ -920,19 +949,41 @@ func (ac *addrConn) resetTransport() error {
 		ac.ready = nil
 	}
 	ac.transport = nil
-	ac.curAddr = resolver.Address{}
+	prevAddr := ac.prevAddr
 	ac.mu.Unlock()
 	ac.cc.mu.RLock()
 	ac.dopts.copts.KeepaliveParams = ac.cc.mkp
 	ac.cc.mu.RUnlock()
+	var (
+		backoffDeadline time.Time
+		connectDeadline time.Time
+	)
 	for retries := 0; ; retries++ {
-		sleepTime := ac.dopts.bs.backoff(retries)
-		timeout := minConnectTimeout
 		ac.mu.Lock()
-		if timeout < time.Duration(int(sleepTime)/len(ac.addrs)) {
-			timeout = time.Duration(int(sleepTime) / len(ac.addrs))
+		if ac.backoffDeadline.IsZero() {
+			// This means either a successfull HTTP2 connection was established
+			// or this is the first time this addrConn is trying to establish a
+			// connection.
+			backoffFor := ac.dopts.bs.backoff(retries) // time.Duration.
+			// This will be the duration that dial gets to finish.
+			dialDuration := minConnectTimeout
+			if backoffFor > dialDuration {
+				// Give dial more time as we keep failing to connect.
+				dialDuration = backoffFor
+			}
+			now := time.Now()
+			backoffDeadline = now.Add(backoffFor)
+			connectDeadline = now.Add(dialDuration)
+			prevAddr = resolver.Address{} // Start connecting from the begining.
+		} else {
+			// Continue trying to conect with the same deadlines.
+			retries = ac.retries
+			backoffDeadline = ac.backoffDeadline
+			connectDeadline = ac.connectDeadline
+			ac.backoffDeadline = time.Time{}
```

Comment:
This is needed since we pass a closure to the transport to update these variables upon receiving settings. Now an error can occur while writing and resetTransport might get invoked while these settings might get updated by that closure.
By doing this, we can prevent that closure from updating these variables.